com.unity.ml-agents.extensions/Runtime/MultiAgent/BaseMultiAgentGroup.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Unity.MLAgents.Extensions.MultiAgent
+{
+    /// <summary>
+    /// A base class implementation of MultiAgentGroup.
+    /// </summary>
+    public class BaseMultiAgentGroup : IMultiAgentGroup, IDisposable
+    {
+        readonly int m_Id = MultiAgentGroupIdCounter.GetGroupId();
+        HashSet<Agent> m_Agents = new HashSet<Agent>();
+
+
+        public void Dispose()
+        {
+            while (m_Agents.Count > 0)
+            {
+                UnregisterAgent(m_Agents.First());
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void RegisterAgent(Agent agent)
+        {
+            if (!m_Agents.Contains(agent))
+            {
+                agent.SetMultiAgentGroup(this);
+                m_Agents.Add(agent);
+                agent.OnAgentDisabled += UnregisterAgent;
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void UnregisterAgent(Agent agent)
+        {
+            if (m_Agents.Contains(agent))
+            {
+                m_Agents.Remove(agent);
+                agent.OnAgentDisabled -= UnregisterAgent;
```

Comment:
I think you need to do something to the Agent here to reset its group ID (and maybe zero out its team reward). You should either make Agent.SetMultiAgentGroup() take `null` and set the ID to 0 in that case, or add Agent.ClearMultiAgentGroup() and do it there.

Reply:
Basically, I think you should have a test like
```csharp
IMultiAgentGroup coach = ...
Agent agent = ....

coach.RegisterAgent(agent);
coach.AddTeamReward(1.0);
coach.RemoveAgent(agent);

Assert.AreEqual(0, agent.groupId);
Assert.AreEqual(0.0f, agent.groupReward); // maybe
```

Reply:
Update accordingly except the group reward part. I kept the group reward reset cycle the same as the individual reward to avoid it being zeroed out before sending to trainer/policy. Also added tests.